File: 2020/old_day14/day14.py
import itertools
from parse import parse

input = '''
mask = 000000000000000000000000000000X1001X
mem[42] = 100
mask = 00000000000000000000000000000000X0XX
mem[26] = 1
'''.strip('\n')

rows = input.split('\n')

mem = {}
mask = ''
mask0 = 0
mask1 = 0
maskX = 0

for row in rows:
  p = parse("mask = {}", row)
  if p:
    mask = p[0]
    mask0 = int(p[0].replace('1', 'X').replace('0', '1').replace('X', '0'), 2)
    mask1 = int(p[0].replace('0', 'X').replace('X', '0'), 2)
    maskX = int(p[0].replace('1', '0').replace('0', '0').replace('X', '1'), 2)
    countXArr = [i for i in range(len(mask)) if mask[i] == 'X']
  else:
    parsed = parse('mem[{:d}] = {:d}', row)
    memIndex, memValue = parsed
    memIndex = (memIndex & ~maskX) | mask1
    for possibility in itertools.product('01', repeat=len(countXArr)):
      v = list(mask)
      for i, j in enumerate(possibility):
        v[countXArr[i]] = j
      addr = int(''.join(v), 2) | memIndex
      mem[addr] = memValue

print(sum(mem.values()))

File: 2020/day14/old/day14.js
console.clear()

console.log('--- JAVASCRIPT ---')

let input = `mask = XXXXXXXXXXXXXXXXXXXXXXXXXXXXX1XXXX0X
mem[8] = 11
mem[7] = 101
mem[8] = 0`.split('\n')


const byte2Dec = (data) => {
  const bigInt = BigInt(`0b${data}`)
  return bigInt
}
const part1 = () => {
  let mem = []
  mask0 = 0
  mask1 = 0

  input.forEach((row, index) => {
    if (row.startsWith('mask = ')) {
      mask0 = byte2Dec(row.replace('mask = ', '').replace(/[1]/g, 'X').replace(/[0]/g, '1').replace(/[X]/g, '0'))
      mask1 = byte2Dec(row.replace('mask = ', '').replace(/[0]/g, 'X').replace(/[X]/g, '0'))
    } else {
      let [_, memIndex, memValue] = row.match(/mem\[(\d{1,})\] = (\d{1,})/)
      memValue = BigInt(memValue)
      memValue &= ~mask0
      memValue |= mask1
      mem[memIndex] = memValue
    }
  })
  return mem
}

const part2 = () => {
  const dec2bin = dec => (dec >>> 0).toString(2)

  let mem = []
  let mask = ''
  let mask0 = 0
  let mask1 = 0
  let maskX = 0
  let countX = 0
  let addr;
  
  input.forEach((row, index) => {
    if (row.startsWith('mask = ')) {
      row = row.replace('mask = ', '')
      mask = row
      mask0 = byte2Dec(row.replace(/[1]/g, 'X').replace(/[0]/g, '1').replace(/[X]/g, '0'))
      mask1 = byte2Dec(row.replace(/[0]/g, 'X').replace(/[X]/g, '0'))
      maskX = byte2Dec(row.replace(/[1]/g, '0').replace(/[0]/g, '0').replace(/[X]/g, '1'))
      countX = row.split('').filter(char => char === 'X')
    } else {
      let [_, memIndex, memValue] = row.match(/mem\[(\d{1,})\] = (\d{1,})/)
      memIndex = BigInt(memIndex)
      memValue = BigInt(memValue)

      memIndex = (memIndex & ~maskX) | mask1

      for(let i = 0; i < countX.length; i++) {
        const lengthOfArray = 2**countX[i].length
        console.log('-----')
        for(let j = 0; j < lengthOfArray; j++) {
          let finalArr = []
          const value = dec2bin(j).split('')
          for(let k = 0; k < countX[i] - value.length; k++) {
            finalArr.push('0')
          }
          value.forEach(sign => {
            finalArr.push(sign)
          });
        
          v = mask.split('')
          finalArr.forEach((j, i) => {
            v[countX[i]] = j
          })
          addr = BigInt([''].join(v), 2) | memIndex
          mem[addr] = memValue
        }
      }
    }
  })
  

  return mem
}

// console.log('------ TASK 1 ------')
const task1 = part1().reduce((curr, prev) => curr + prev)
console.log(task1)
console.log(parseInt(task1, 10))
// console.log('------- END1 -------\n')

// console.log('------ TASK 2 ------')
// const task2 = part2().reduce((curr, prev) => curr + prev)
// console.log(task2)
// console.log('------- END2 -------\n')